resources/migrations/001_update_migrations.yaml

```diff
@@ -8253,6 +8253,151 @@ databaseChangeLog:
+      changes:
+        - addColumn:
+            tableName: report_dashboard
+            columns:
+              - column:
+                  name: last_viewed_at
+                  type: ${timestamp_type}
+                  remarks: Timestamp of when this dashboard was last viewed
+                  constraints:
+                    nullable: true
+
+  - changeSet:
+      id: v51.2024-06-14T06:45:34
+      author: johnswanson
+      comment: Create index on view_log (model, model_id, timestamp)
+      changes:
+        - createIndex:
+            tableName: view_log
+            columns:
+              - column:
+                  name: model
+              - column:
+                  name: model_id
+              - column:
+                  name: timestamp
+            indexName: idx_view_log_model_model_id_timestamp
+
+  - changeSet:
+      id: v51.2024-06-14T06:45:35
+      author: johnswanson
+      comment: Populate `last_viewed_at` for dashboards
+      rollback: # not needed, will be removed
+      changes:
+        - sql:
+            dbms: mariadb,mysql
+            sql: >-
+              UPDATE report_dashboard rd
+              LEFT JOIN (
+                  SELECT dashboard_id, MAX(COALESCE(last_used_at, '1970-01-01 00:00:00')) AS max_card_last_used
+                  FROM report_dashboardcard
+                  JOIN report_card ON report_dashboardcard.card_id = report_card.id
+                  GROUP BY dashboard_id
+              ) card_data ON rd.id = card_data.dashboard_id
+              LEFT JOIN (
+                  SELECT model_id, MAX(timestamp) AS max_view_timestamp
+                  FROM view_log
+                  WHERE model = 'dashboard'
+                  GROUP BY model_id
+              ) view_data ON rd.id = view_data.model_id
+              SET rd.last_viewed_at = COALESCE(
+                  NULLIF(
+                      GREATEST(
+                          COALESCE(card_data.max_card_last_used, '1970-01-01 00:00:00'),
+                          COALESCE(view_data.max_view_timestamp, '1970-01-01 00:00:00')
+                      ),
+                      '1970-01-01 00:00:00'
+                  ),
+                  NOW()
+              );
+        - sql:
+            dbms: h2
+            sql: >-
+              UPDATE report_dashboard
+              SET last_viewed_at = COALESCE(GREATEST(
+                (
+                  SELECT MAX(timestamp)
+                  FROM view_log
+                  WHERE report_dashboard.id = model_id
+                    AND model = 'dashboard'
+                ),
+                (
+                   SELECT
+                     MAX(last_used_at) AS last_card_used_at
+                   FROM report_card
+                   JOIN report_dashboardcard
+                     ON report_dashboardcard.card_id = report_card.id
+                   WHERE report_dashboardcard.dashboard_id = report_dashboard.id
+                )),
+                NOW())
```

Comment:
This would be awesome if it worked but unfortunately taking `greatest(max(report_card.last_used_at), max(view_log.timestamp))` can result in some `last_viewed_at` values that are further back in the past than the dashboard was actually viewed.

One problem is `view_log` used to be populated in OSS instances, but we removed it from OSS in v50. And `last_used_at` was [pulled out from some versions of v50](https://github.com/metabase/metabase/pull/44502). That means this scenario is possible:
1. A instance is created on version 49, using OSS
2. Dashboard A was viewed on date X, creating a `view_log` entry. (There's no `last_used_at` because `report_card.last_used_at` was introduced in 50.)
3. The instance upgrades to version 50.7 where `last_used_at` is not populated.
4. Dashboard A is viewed again on date Y. No `view_log` entry is created or `last_used_at` and no `last_used_at` value is set.
5. The instance upgrades to version 51.0. The `report_dashboard.last_viewed_at` value for dashboard A is X, but the logically correct value is Y.

There's more scenarios too, which might be more likely:
1. An EE instance is on version 50.7 or any other version in 50 where `last_used_at` is not present but `view_log` is being added to.
2. Dashboard A was viewed on date X, creating a `view_log` entry.
3. The instance changes from EE to OSS.
4.  Dashboard A is viewed again on date Y. No `view_log` entry is created because it's OSS or `last_used_at` and no `last_used_at` value is set.
5. The instance upgrades to version 51.0. The `report_dashboard.last_viewed_at` value for dashboard A is X, but the logically correct value is Y.

There's also another scenario where an OSS instance upgrades from 50.6 to 50.7, and no `last_used_at`. I could go on.

I can't see a fool-proof workaround to these problems :( And because this will be used for auto-clean-up I think we'd prefer to be conservative. To avoid setting `last_used_at` further in the past than reality in all scenarios, I think it has to default to `now`.

Reply:
This would certainly make the migrations a lot easier and lower risk 😅

Reply:
@calherries I realized that there's a more general issue here.

1. Say someone is currently on v50.6 (populating `report_card.last_used_at`)
2. They upgrade tomorrow to v50.7 (we stop populating the column)
3. In 7 months, they upgrade to v51 and we release the "Show me my stale items" feature

We'll happily tell them "all of your cards are stale!" because they have non-null `last_used_at` values from >6 months ago. I'm thinking about what our options are here.

Reply:
Yes exactly, that's the scenario I (loosely) described here:

> There's also another scenario where an OSS instance upgrades from 50.6 to 50.7, and last_used_at goes from being updated to not.